apollo-compiler/src/test/graphql/com/example/fragment_in_fragment/AllStarships.kt

```diff
@@ -86,57 +89,381 @@ class AllStarships : Query<AllStarships.Data, AllStarships.Data, Operation.Varia
+   * 0 ABY.
+   */
+  interface Homeworld : PilotFragment.Homeworld {
+    override val __typename: String
+
+    override fun marshaller(): ResponseFieldMarshaller
+
+    companion object {
+      private val RESPONSE_FIELDS: Array<ResponseField> = arrayOf(
+        ResponseField.forString("__typename", "__typename", null, false, null)
+      )
+
+      operator fun invoke(reader: ResponseReader): Homeworld {
+        val typename = reader.readString(RESPONSE_FIELDS[0])
+        return when(typename) {
+          "Planet" -> PlanetFragmentImpl(reader)
+          else -> HomeworldImpl(reader)
+        }
+      }
+    }
+  }
+
+  /**
+   * An individual person or character within the Star Wars universe.
+   */
+  data class PilotFragmentImpl(
+    /**
+     * The name of this person.
+     */
+    override val name: String?,
+    /**
+     * A planet that this person was born on or inhabits.
+     */
+    override val homeworld: Homeworld?,
+    override val __typename: String = "Person"
+  ) : Node1, PilotFragment, StarshipFragment.Node {
+    override fun marshaller(): ResponseFieldMarshaller {
+      return ResponseFieldMarshaller.invoke { writer ->
+        writer.writeString(RESPONSE_FIELDS[0], this@PilotFragmentImpl.name)
+        writer.writeObject(RESPONSE_FIELDS[1], this@PilotFragmentImpl.homeworld?.marshaller())
+        writer.writeString(RESPONSE_FIELDS[2], this@PilotFragmentImpl.__typename)
+      }
+    }
+
+    companion object {
+      private val RESPONSE_FIELDS: Array<ResponseField> = arrayOf(
+        ResponseField.forString("name", "name", null, true, null),
+        ResponseField.forObject("homeworld", "homeworld", null, true, null),
+        ResponseField.forString("__typename", "__typename", null, false, null)
+      )
+
+      operator fun invoke(reader: ResponseReader): PilotFragmentImpl = reader.run {
+        val name = readString(RESPONSE_FIELDS[0])
+        val homeworld = readObject<Homeworld>(RESPONSE_FIELDS[1]) { reader ->
+          Homeworld(reader)
+        }
+        val __typename = readString(RESPONSE_FIELDS[2])!!
+        PilotFragmentImpl(
+          name = name,
+          homeworld = homeworld,
+          __typename = __typename
+        )
+      }
+
+      @Suppress("FunctionName")
+      fun Mapper(): ResponseFieldMapper<PilotFragmentImpl> = ResponseFieldMapper { invoke(it) }
+    }
+  }
+
+  /**
+   * An individual person or character within the Star Wars universe.
+   */
+  interface Node1 : StarshipFragment.Node {
+    override val __typename: String
+
+    override fun marshaller(): ResponseFieldMarshaller
+
+    companion object {
+      private val RESPONSE_FIELDS: Array<ResponseField> = arrayOf(
+        ResponseField.forString("__typename", "__typename", null, false, null)
+      )
+
+      operator fun invoke(reader: ResponseReader): Node1 {
+        val typename = reader.readString(RESPONSE_FIELDS[0])
+        return when(typename) {
+          "Person" -> PilotFragmentImpl(reader)
```

Comment:
Feature Request: if we know the type condition is always true, could we implement the interface by default? 

```
interface Node1: StarshipFragment.Node, PilotFragment
```

Reply:
Could you pls elaborate on this, not sure I understand this feature request.

Reply:
Yes, here `starship.pilotConnection.edges[0].node` is of GraphQL type `Person` but the generated type is `Node1`, meaning that users will always have to cast to access pilot fields while we know at compile time that they should always be there.

I'm not 100% sure how to generate this code but I feel like making "Node1" implement "PilotFragment" when Node1 is always of type "Person" would be handy. Of course if there are more possible types it wouldn't work anymore but I feel like the usage could be simplified for this specific case.

Reply:
Agh right, nice catch. Now I have to think how to formalize this optimization in terms of codegen.

Reply:
Done